src/hotspot/share/runtime/threadSMR.inline.hpp

```diff
@@ -33,6 +33,59 @@
 #include "memory/iterator.hpp"
 #include "runtime/prefetch.inline.hpp"
 #include "runtime/thread.inline.hpp"
+#include "utilities/debug.hpp"
+#include "utilities/macros.hpp"
+
+ThreadsList::Iterator::Iterator(ThreadsList* list, uint i) :
+  _thread_ptr(list->threads() + check_index(list, i))
+  DEBUG_ONLY(COMMA _list(list))
+{}
+
+bool ThreadsList::Iterator::operator==(Iterator i) const {
+  assert_not_singular();
+  assert_same_list(i);
+  return _thread_ptr == i._thread_ptr;
+}
+
+bool ThreadsList::Iterator::operator!=(Iterator i) const {
+  assert_not_singular();
+  assert_same_list(i);
+  return _thread_ptr != i._thread_ptr;
+}
+
+JavaThread* ThreadsList::Iterator::operator*() const {
+  assert_not_singular();
+  assert_dereferenceable();
+  Prefetch::read(const_cast<JavaThread**>(_thread_ptr), PrefetchScanIntervalInBytes);
```

Comment:
This prefetch is included because the old code had it.  But I'm dubious about it.  This is just linear iteration through an array, which seems like the canonical best case for leaving it to the hardware, rather than doing cachelinesize/ptrsize software prefetches per cache line.  I'm hoping the original authors can comment.

Reply:
The original author that added the prefetch is @fisk. He's reviewed an earlier
version of this fix and I'm hoping he's around to review this version.

Reply:
I agree that it should be okay without prefetch. I added it when I was running on a 4096 strand SPARC machine running over 8000 threads, and walking the list would approach 1 ms. While intuitively linear scan should be good enough, I did actually measure slight improvements with explicit prefetching, and the slight improvements did seem worth it at the time as the amount of time spent in there was non-trivial.
Now we don't support that kind of hardware any longer, so I don't think this trick would necessarily spark as much joy any longer.
I don't mind if we do or do not perform the prefetching. I'm okay either way.

Reply:
I'm going to leave resolution of the prefetch question to a follow-up bug.
I think that discussion deserves a separate issue.

Reply:
OK.